### ISSUE TITLE
[data grid] Limit number of checkboxes selected in data grid

### ISSUE THREAD

**Issue by tvaughan73** [mui/mui-x#9720]
### Order ID

62910

### The problem in depth 🔍

When using the data grid, how can I limit the number of rows that can be selected with a checkbox?  
What I would like to happen is when five checkboxes are checked all other checkboxes would become disabled.  Is there a way to do this?

I searched and found https://github.com/mui/mui-x/issues/5241 but the solution does not work.  

I am already using this to listen for row selection change and I thought it may be able to be used for this purpose as well.

    useEffect(() => {
        const handleEvent: GridEventListener<"rowSelectionChange"> = (
            params,
        ) => {
            if (apiRef.current.getSelectedRows().size > 0) {
                setShowDownloadButton(true);
            } else {
                setShowDownloadButton(false);
            }
        };
        return apiRef.current.subscribeEvent("rowSelectionChange", handleEvent);
    }, []);

### Your environment 🌎

<details>
  <summary>`npx @mui/envinfo`</summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by romgrk** [comment]
Have you looked at the documentation for [controlled row selection](https://mui.com/x/react-data-grid/row-selection/#controlled-row-selection)?

**Comment by tvaughan73** [comment]
Yes, I have.  Here is a sandbox using both.  The problem, at least how this is coded, using isRowSelectable does not produce the desired behavior.    Check a few rows to see what I mean.  And then try to deselect.

https://codesandbox.io/s/loving-ptolemy-qfjx3x?file=/demo.tsx

**Comment by romgrk** [comment]
The logic is more correct like this: https://codesandbox.io/s/agitated-meadow-rj36fl

The selectable state still lags behind in some cases, we should investigate further.

**Comment by tvaughan73** [comment]
Thanks!

**Comment by michelengelen** [comment]
I just retested this and could not notice any more lagging.
The adjusted demo code for this (`GridRowSelectionModel` structure changed since this was opened) is this:

```tsx
import * as React from "react";
import {
  DataGrid,
  GridRowParams,
  GridRowSelectionModel,
  useGridApiRef
} from "@mui/x-data-grid";
import { useDemoData } from "@mui/x-data-grid-generator";

export default function DisableRowSelection() {
  const apiRef = useGridApiRef();
  const { data } = useDemoData({
    dataSet: "Commodity",
    rowLength: 100,
    maxColumns: 6
  });

  const [rowSelectionModel, setRowSelectionModel] = React.useState<
    GridRowSelectionModel
  >({ ids: new Set(), type: 'include'});

  const isRowSelectable = React.useCallback(
    (params: GridRowParams) => {
      if (rowSelectionModel.ids.has(params.id)) return true;
      return rowSelectionModel.ids.size < 2;
    },
    [rowSelectionModel]
  );

  return (
    <div style={{ height: 400, width: "100%" }}>
      <DataGrid
        sx={{
          "& .MuiDataGrid-columnHeaderCheckbox .MuiDataGrid-columnHeaderTitleContainer": {
            display: "none"
          }
        }}
        {...data}
        apiRef={apiRef}
        isRowSelectable={isRowSelectable}
        checkboxSelection
        onRowSelectionModelChange={setRowSelectionModel}
        rowSelectionModel={rowSelectionModel}
      />
    </div>
  );
}
```

if you still experience issues with this please feel free to open a new issue! Thanks!
